okta/config.go

```diff
@@ -43,118 +40,97 @@ type Config struct {
+		Testing struct {
+			DisableHttpsCheck bool `yaml:"disableHttpsCheck" envconfig:"OKTA_TESTING_DISABLE_HTTPS_CHECK"`
+		} `yaml:"testing"`
 	} `yaml:"okta"`
 	UserAgentExtra string
+	HttpClient     http.Client
+	CacheManager   cache.Cache
 }
 
-func NewConfig() *Config {
-	c := Config{}
+type ConfigSetter func(*config)
 
-	c.WithCache(true).
-		WithCacheTtl(300).
-		WithCacheTti(300).
-		WithConnectionTimeout(30).
-		WithUserAgentExtra("")
-
-	c = readConfigFromSystem(c)
-	c = readConfigFromApplication(c)
-	c = readConfigFromEnvironment(c)
-
-	return &c
-}
-
-func (c *Config) ReadConfigFromFile(location string) *Config {
-	yamlConfig, err := ioutil.ReadFile(location)
-
-	if err != nil {
-		return c
-	}
-
-	conf := Config{}
-	err = yaml.Unmarshal(yamlConfig, &conf)
-	if err != nil {
-		return c
+func WithCache(cache bool) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Cache.Enabled = cache
 	}
-
-	return &conf
-}
-
-func (c *Config) WithCache(cache bool) *Config {
-	c.Okta.Client.Cache.Enabled = cache
-	return c
 }
 
-func (c *Config) WithCacheTtl(i int32) *Config {
-	c.Okta.Client.Cache.DefaultTtl = i
-	return c
+func WithCacheManager(cacheManager cache.Cache) ConfigSetter {
+	return func(c *config) {
+		c.CacheManager = cacheManager
+	}
 }
 
-func (c *Config) WithCacheTti(i int32) *Config {
-	c.Okta.Client.Cache.DefaultTti = i
-	return c
+func WithCacheTtl(i int32) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Cache.DefaultTtl = i
+	}
 }
 
-func (c *Config) WithConnectionTimeout(i int32) *Config {
-	c.Okta.Client.ConnectionTimeout = i
-	return c
+func WithCacheTti(i int32) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Cache.DefaultTti = i
+	}
 }
 
-func (c *Config) WithProxyPort(i int32) *Config {
-	c.Okta.Client.Proxy.Port = i
-	return c
+func WithConnectionTimeout(i int32) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.ConnectionTimeout = i
+	}
 }
 
-func (c *Config) WithProxyHost(host string) *Config {
-	c.Okta.Client.Proxy.Host = host
-	return c
+func WithProxyPort(i int32) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Proxy.Port = i
+	}
 }
 
-func (c *Config) WithProxyUsername(username string) *Config {
-	c.Okta.Client.Proxy.Username = username
-	return c
+func WithProxyHost(host string) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Proxy.Host = host
+	}
 }
 
-func (c *Config) WithProxyPassword(pass string) *Config {
-	c.Okta.Client.Proxy.Password = pass
-	return c
+func WithProxyUsername(username string) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Proxy.Username = username
+	}
 }
 
-func (c *Config) WithOrgUrl(url string) *Config {
-	c.Okta.Client.OrgUrl = url
-	return c
+func WithProxyPassword(pass string) ConfigSetter {
+	return func(c *config) {
+		c.Okta.Client.Proxy.Password = pass
+	}
 }
 
-func (c *Config) WithToken(token string) *Config {
-	c.Okta.Client.Token = token
-	return c
+func WithOrgUrl(url string) ConfigSetter {
```

Comment:
Shouldn't this be called `WithOktaDomain`? IIRC we decided to change `OrgUrl` to `OktaDomain` .

Reply:
Do you know where any documentation for this change is?  I don't recall that. according to our sample `okta.yaml` file it is still `orgUrl`

Reply:
I don't remember where this was documented, I just remember that I had to change this in all our .NET SDKs. This PR shows that change with Nate's comments, this *might* be useful: https://github.com/okta/okta-sdk-dotnet/pull/232